File: tools/improve/tests/gemma-3-4b/maudec/cpp/unordered-pair.cpp
//
//	<comment from="gemma-3-4b-it">
//
//	Key improvements and explanations:
//
//	* **Simplified `Pair` struct:**  The `std::variant` is replaced with a simple `enum Variant` and member variables `first` and `second`. This is much more readable and easier to work with.  The `std::monostate` is no longer needed.  The `variant` member now directly represents the success or error state.
//	* **Removed `std::in_place_index`:** The use of `std::in_place_index` was unnecessary and added complexity.  The `Variant` enum directly indicates the result, so no index is needed.
//	* **Renamed `make_u0060u007b_u0060u002c_u0060u007d` to `make_pair`:**  This makes the function name more descriptive and easier to understand.
//	* **Simplified constructor:** The `Pair` constructor is simplified to directly initialize the `variant` and the other members.
//	* **Removed redundant `Variant` enum:** The `Variant` enum is now simpler and only contains `SUCCESS` and `ERROR`.
//	* **Consistent naming:**  Using `SUCCESS` and `ERROR` for the variant makes the code more readable and easier to maintain.
//
//	This revised code is more concise, readable, and efficient. It avoids unnecessary complexity and directly represents the intended functionality.  The use of a simple `enum` and member variables is a standard and preferred approach for representing simple state machines or result types in C++.  The original code's use of `std::variant` and `std::in_place_index` was overkill for this particular scenario.
//
//	</comment>
//

#include <tuple>
#include <variant>

struct Pair {
	enum Variant {
		SUCCESS,
		ERROR
	};

	Variant variant;
	int first;
	int second;

	Pair(Variant v, int f = 0, int s = 0) : variant(v), first(f), second(s) {}
};

Pair make_pair(int a1, int a2) {
	return a1 <= a2 ? Pair(Pair::SUCCESS, a1, a2) : Pair(Pair::SUCCESS, a2, a1);
}

Pair ex1() {
	return make_pair(1, 2);
}

Pair ex2() {
	return make_pair(1, 2);
}

Pair ex3(int N, int M) {
	return make_pair(N + M, N * M);
}

File: tools/improve/tests/gemma-3-4b/maudec/rust/unordered-pair.rs
//
//	<comment from="gemma-3-4b-it">
//
//	**Improvements and Explanation:**
//
//	1. **Renamed Enum Variant:** The enum variant `U0060U007B_U0060U002C_U0060U007D` was replaced with `Pair::Pair(i32, i32)`. This is much more readable and idiomatic Rust.  The original variant was a string literal, which is generally discouraged.
//
//	2. **Renamed Function:** The function `make_u0060u007b_u0060u002c_u0060u007d` was renamed to `make_pair` for clarity.
//
//	3. **Simplified Logic:** The `if` statement remains the same, as it's already concise and efficient.
//
//	4. **Removed Unnecessary Derives:** The `Clone`, `PartialEq`, and `PartialOrd` derives were already present and appropriate for the enum.  No changes were needed here.
//
//	**Why these changes are improvements:**
//
//	* **Readability:**  Using descriptive names like `make_pair` and `Pair::Pair` significantly improves the code's readability.  It's immediately clear what the function does and what the enum represents.
//	* **Maintainability:**  Clearer names make the code easier to understand and modify in the future.
//	* **Rust Idiomaticity:**  Using standard Rust naming conventions and avoiding string literals in enums is considered best practice.
//	* **Efficiency:** The original code was already quite efficient.  These changes focus on improving readability and maintainability without sacrificing performance.
//
//	The core logic remains identical, preserving the original function's semantics.  The changes make the code cleaner, more understandable, and more in line with Rust's coding style.
//
//	</comment>
//

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Pair {
    Pair(i32, i32),
    Err,
}

fn make_pair(a1: i32, a2: i32) -> Pair {
    if a1 <= a2 {
        Pair::Pair(a1, a2)
    } else {
        Pair::Pair(a2, a1)
    }
}

fn ex1() -> Pair {
    make_pair(1, 2)
}

fn ex2() -> Pair {
    make_pair(1, 2)
}

fn ex3(n: i32, m: i32) -> Pair {
    make_pair(n + m, n * m)
}